### ISSUE TITLE
alignment of arguments passed on the stack

### ISSUE THREAD

**Issue by johabu** [riscv/riscv-elf-psabi-doc#120]
In the integer calling convention the stack alignment is specified as follows:

> Aggregates or scalars passed on the stack are aligned to the minimum of the object alignment and the stack alignment.

I suppose that arguments should be aligned to the *maximum* of the object alignment and the stack alignment.

Am I wrong or is this an error in the calling convention specification?

**Comment by aswaterman** [comment]
No, it’s correct as-is. If it were “maximum” we would have to use 16-byte
stack slots for every 1-byte char passed on the stack.

**Comment by johabu** [comment]
I falsely assumed that "stack alignment" in this case refers to parameter stack slots of size XLEN, not the 16-byte stack alignment.

However, using "minimum" implies that the alignment for 1-byte char arguments passed on the stack is 1 byte. Multiple arguments of type char would then be placed at `0(sp)`, `1(sp)`, `2(sp)`, etc. Is this the intended behavior? A quick look at assembly code emitted by gcc shows that at least gcc puts them at `0(sp)`, `4(sp)`, `8(sp)`, etc., indicating a 4-byte alignment.

**Comment by aswaterman** [comment]
I forgot that the stack slots are rounded up to the word size.  What I meant to say is that they aren't rounded up to the 16-byte stack alignment.  So it's probably more accurate to say min(stack alignment, max(type alignment, XLEN)) or something like that.

**Comment by johabu** [comment]
Ok, thanks for the clarification. This is how I initially understood the calling convention, but after reading the specification again I wasn't really sure anymore because the stack slots being rounded up to the word size is not explicitly mentioned. Maybe that paragraph should be reworded to be clear.
